### PR TITLE
o/snapstate: mk pre-download tasks if running snap blocks autorefresh

### PR DESCRIPTION
**Pull request by MiguelPires**
If an auto-refresh is totally or partially prevented due to some snaps being open, create pre-download tasks so that, when the user is asked to close the snap, it's ready to be auto-refreshed as quickly as possible. If the inhibition window has passed, pre-downloads aren't created and instead the user is notified that the refresh is proceeding despite the open snap. The next PR will add the actual pre-download logic as well as the related logic w.r.t waiting for snaps to close, concurrent downloads, continuing delayed auto-refreshes, etc.

### CODE REVIEW

File: overlord/snapstate/snapstate.go [hunk 1]
```diff
@@ -665,6 +685,26 @@ func doInstall(st *state.State, snapst *SnapState, snapsup *SnapSetup, flags int
 	return ts, nil
 }
 
+func findTasksMatching(st *state.State, kind string, snapName string, revision snap.Revision) ([]*state.Task, error) {
```
Comment (pedronis) [inline overlord/snapstate/snapstate.go:688]: should this be called findTasksMatchingKindAndSnap ?

**Review by pedronis — COMMENTED** [review]
thank you, did a pass, some questions and it seems we are missing some unit tests

File: overlord/snapstate/handlers.go [hunk 2]
```diff
@@ -3706,7 +3706,7 @@ func (m *SnapManager) doCheckReRefresh(t *state.Task, tomb *tomb.Tomb) error {
 		return err
 	}
 
-	updated, tasksets, err := reRefreshUpdateMany(tomb.Context(nil), st, snaps, nil, re.UserID, reRefreshFilter, re.Flags, chg.ID())
+	updated, tasksetGroup, err := reRefreshUpdateMany(tomb.Context(nil), st, snaps, nil, re.UserID, reRefreshFilter, re.Flags, chg.ID())
```
[the remark this answers is not in the document]
Comment (MiguelPires) [reply overlord/snapstate/handlers.go:3709]: It can, I didn't create a pre-download change because at this point we can only re-refresh snaps that have just been refreshed so it would be very unlikely that the snap becomes busy in the meantime. But I guess there's no harm in doing that

**Comment by MiguelPires** [comment]
Thanks for the review @pedronis. Sorry for the missing tests, I forgot to move some when splitting this out of the draft PR. I've corrected that now and also changed the rerefresh corner case you mentioned so it can also create pre-downloads as well

**Review by pedronis — APPROVED** [review]
thank you

this and #12558 might conflict?

File: overlord/snapstate/snapstate.go [hunk 3]
```diff
@@ -1690,12 +1739,23 @@ func filterHeldSnaps(st *state.State, updates []minimalInstallInfo, flags *Flags
 	return filteredUpdates, nil
 }
 
-func doUpdate(ctx context.Context, st *state.State, names []string, updates []minimalInstallInfo, params updateParamsFunc, userID int, globalFlags *Flags, deviceCtx DeviceContext, fromChange string) ([]string, []*state.TaskSet, error) {
+// UpdateTaskSets distinguishes tasksets for refreshes and pre-downloads since an
```
Comment (mvo5) [inline overlord/snapstate/snapstate.go:1742]: Thanks for this comment

File: overlord/snapstate/snapstate.go [hunk 4]
```diff
@@ -2272,11 +2343,14 @@ func UpdateWithDeviceContext(st *state.State, name string, opts *RevisionOptions
 		return opts, flags, &snapst
 	}
 
-	_, tts, err := doUpdate(context.TODO(), st, []string{name}, toUpdate, params, userID, &flags, deviceCtx, fromChange)
+	_, updateTss, err := doUpdate(context.TODO(), st, []string{name}, toUpdate, params, userID, &flags, deviceCtx, fromChange)
 	if err != nil {
 		return nil, err
 	}
 
+	// pre-download tasksets are only used for auto-refreshes
```
Comment (mvo5) [inline overlord/snapstate/snapstate.go:2351]: Maybe expand the comment slightly: `// pre-download tasksets are only used for auto-refreshes so they are ignored here`  (unless I misunstand the code?

File: overlord/snapstate/snapstate.go [hunk 5]
```diff
@@ -2581,13 +2660,17 @@ func autoRefreshPhase2(ctx context.Context, st *state.State, updates []*refreshC
...
 
-	updated, tasksets, err := doUpdate(ctx, st, nil, toUpdate, nil, userID, flags, deviceCtx, fromChange)
+	updated, updateTss, err := doUpdate(ctx, st, nil, toUpdate, nil, userID, flags, deviceCtx, fromChange)
 	if err != nil {
 		return nil, err
 	}
 
-	tasksets = finalizeUpdate(st, tasksets, len(updates) > 0, updated, userID, flags)
-	return tasksets, nil
+	// pre-download tasksets are only used for auto-refreshes
```
Comment (mvo5) [inline overlord/snapstate/snapstate.go:2668]: (same comment about "so they are ignored here" as above maybe?)

**Review by mvo5 — APPROVED** [review]
Thanks, this looks very nice. I have tiny suggestions inline, feel free to ignore or address in followups

File: overlord/snapstate/snapstate.go (hunk 4 above)
Comment (MiguelPires) [reply overlord/snapstate/snapstate.go]: Done, thank you for the review
